nexus/db-queries/src/db/datastore/ipv4_nat_entry.rs

```diff
@@ -0,0 +1,495 @@
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupResult;
+use omicron_common::api::external::LookupType;
+use omicron_common::api::external::ResourceType;
+
+impl DataStore {
+    pub async fn ensure_ipv4_nat_entry(
+        &self,
+        opctx: &OpContext,
+        nat_entry: Ipv4NatValues,
+    ) -> CreateResult<()> {
+        // Ensure that the record with the parameters we want exists in the
+        // database. If an entry exists with the same external ip but
+        // different target sled / vpc / mac, we will violate a uniqueness
+        // constraint (which is desired in this case).
+        //
+        // It seems that this isn't straightforward to express in diesel
+        sql_query(
+            "
+            INSERT INTO omicron.public.ipv4_nat_entry (
+                external_address,
+                first_port,
+                last_port,
+                sled_address,
+                vni,
+                mac
+            )
+            SELECT external_address, first_port, last_port, sled_address, vni, mac
+            FROM ( VALUES ($1, $2, $3, $4, $5, $6) ) AS data (external_address, first_port, last_port, sled_address, vni, mac)
+            WHERE NOT EXISTS (
+                SELECT external_address, first_port, last_port, sled_address, vni, mac
+                FROM omicron.public.ipv4_nat_entry as entry
+                WHERE data.external_address = entry.external_address
+                AND data.first_port = entry.first_port
+                AND data.last_port = entry.last_port
+                AND data.sled_address = entry.sled_address
+                AND data.vni = entry.vni
+                AND data.mac = entry.mac
+            )
+            ",
+        )
+            .bind::<Inet, _>(nat_entry.external_address)
+            .bind::<Integer, _>(nat_entry.first_port)
+            .bind::<Integer, _>(nat_entry.last_port)
+            .bind::<Inet, _>(nat_entry.sled_address)
+            .bind::<Integer, _>(nat_entry.vni)
+            .bind::<BigInt, _>(nat_entry.mac)
+            .execute_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(())
+    }
+
+    pub async fn ipv4_nat_delete(
+        &self,
+        opctx: &OpContext,
+        nat_entry: &Ipv4NatEntry,
+    ) -> DeleteResult {
+        // We use pure SQL here so we can call nextval() to increment the sequence number in the
+        // database.
+        let updated_rows = sql_query(
+            "
+            UPDATE omicron.public.ipv4_nat_entry
+            SET
+            version_removed = nextval('omicron.public.ipv4_nat_version'),
+            time_deleted = now()
+            WHERE time_deleted IS NULL AND version_removed IS NULL AND id = $1 AND version_added = $2
+            ",
+        )
```

Comment:
I think we definitely need custom SQL to handle `nextval`, but the rest of this could be vanilla diesel.

If we add the following function:

```rust
fn ipv4_nat_next_version() -> diesel::expression::SqlLiteral<BigInt> {                                                                                                                                                                                                                                                                                            
    diesel::dsl::sql::<BigInt>("nextval('omicron.public.ipv4_nat_version')")                                                                                                                                                                                                                                                                                      
}
```

Then this particular snippet of SQL could be replaced with the following:

```rust
        use db::schema::ipv4_nat_entry::dsl;                                                                                                                                                                                                                                                                                                                      
        let updated_rows = diesel::update(dsl::ipv4_nat_entry)                                                                                                                                                                                                                                                                                                    
            .set((                                                                                                                                                                                                                                                                                                                                                
                dsl::version_removed.eq(ipv4_nat_next_version().nullable()),                                                                                                                                                                                                                                                                                      
                dsl::time_deleted.eq(Utc::now()),                                                                                                                                                                                                                                                                                                                 
            ))                                                                                                                                                                                                                                                                                                                                                    
            .filter(dsl::time_deleted.is_null())                                                                                                                                                                                                                                                                                                                  
            .filter(dsl::version_removed.is_null())                                                                                                                                                                                                                                                                                                               
            .filter(dsl::id.eq(nat_entry.id))                                                                                                                                                                                                                                                                                                                     
            .filter(dsl::version_added.eq(nat_entry.version_added))                                                                                                                                                                                                                                                                                               
            .execute_async(&*self.pool_connection_authorized(opctx).await?)                                                                                                                                                                                                                                                                                       
            .await                                                                                                                                                                                                                                                                                                                                                
            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;  
```